software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+        # generate output filename
+        outputFile = f"saved_state_{self.__class__.__qualname__}_{self.bankToSave}.txt"
+
+        # Convert each value to an int by multiplying by 100. This saves of storage and memory a little
+        i=0
+        for channel in self.CVR[self.bankToSave]:
+            self.CVR[self.bankToSave][i] = [int(x * 100) for x in self.CVR[self.bankToSave][i]]
+            i += 1
+
+        if self.debugTest:
+            print('Saving state for bank: ' + str(self.bankToSave))
+
+        # Trigger garbage collection to minimize memory use
+        gc.collect()
+
+        # Show free memory if running a debug test
+        if self.debugTest:
+            print(self.free())
+
+        # Write the value to a the state file
+        maxRetries = 6
+        attempts = 0
+        while attempts < maxRetries:
+            try:
+                attempts += 1
+                # Create json object of current CV bank
+                jsonState = json.dumps(self.CVR[self.bankToSave])
+                with open(outputFile, 'w') as file:
+                    # Attempt write data to state on disk, then break from while loop if the return (num bytes written) > 0
+                    if file.write(jsonState) > 0:
+                        file.close()
+                        break
+            except MemoryError as e:
+                if self.debugTest:
+                    print(f'[{attempts}] Error: Memory allocation failed, retrying: {e}')
+                    print(micropython.mem_info("level"))
+                else:
+                    pass
+
+        # Convert from int back to float
+        i=0
+        for channel in self.CVR[self.bankToSave]:
+            self.CVR[self.bankToSave][i] = [x / 100 for x in self.CVR[self.bankToSave][i]]
+            i += 1
+
+    def loadState(self):
+
+        # For each bank, check if a state file exists, then load it
+        # If not, initialize the bank with zeros
+
+        self.CVR = []  # CV recorder channels
+        self.CvRecording = []  # CV recorder flags
+
+        # init cvRecording list
+        for i in range(self.numCVR+1):
+            self.CvRecording.append('false')
+
+        for b in range(self.numCVRBanks+1):
+            # Create a new array for the bank
+            self.CVR.append([])
+
+            # Check if a state file exists
+            fileName = f"saved_state_{self.__class__.__qualname__}_{b}.txt"
+            try:
+                # save state exists for this bank, load it
+                with open(fileName, 'r') as file:
+                    if self.debugTest:
+                        print('Loading previous state for bank: ' + str(b))
+
+                    # read state from file into json object
+                    jsonData = file.read()
+
+                    # populate CV recording channel with saved json data
+                    self.CVR[b] = json.loads(jsonData)
+
+                    # convert values in the list from int back to float
```

Comment:
You do this in two different places. Consider extracting a method.

Reply:
I am happy with this being duplicated just twice for the time being. If I end up doing it more than twice I will definitely move it to a method/function.